File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/utils/Log4Error.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.dllib.utils

import org.apache.logging.log4j.{LogManager, Logger}

object Log4Error {
  val logger: Logger = LogManager.getLogger(getClass)

  def invalidInputError(condition: Boolean, errmsg: String, fixmsg: String = null): Unit = {
    if (!condition) {
      outputUserMessage(errmsg, fixmsg)
      // scalastyle:off
      throw new IllegalArgumentException(errmsg)
      // scalastyle:on
    }
  }

  def invalidOperationError(condition: Boolean, errmsg: String, fixmsg: String = null,
                            cause: Throwable = null): Unit = {
    if (!condition) {
      outputUserMessage(errmsg, fixmsg)
      // scalastyle:off
      throw new InvalidOperationException(errmsg, cause)
      // scalastyle:on
    }
  }

  def outputUserMessage(errmsg: String = null, fixmsg: String = null): Unit = {
    if (errmsg != null) {
      logger.info(s"\n\n****************************Usage Error************************\n"
        + errmsg)
    }
    if (fixmsg != null) {
      logger.info(s"\n\n****************************How to fix*************************\n"
        + fixmsg)
    }
    logger.info(s"\n\n****************************Call Stack*************************")
  }

  def unKnowExceptionError(condition: Boolean, errmsg: String = null,
    fixmsg: String = null, cause: Throwable = null): Unit = {
    if (!condition) {
      outputUserMessage(errmsg, fixmsg)
      // scalastyle:off
      throw new UnKnownException(errmsg, cause)
      // scalastyle:on
    }
  }
}

class InvalidOperationException(message: String, cause: Throwable = null)
  extends Exception(message, cause) {

  def this(message: String) = this(message, null)
}

class UnKnownException(message: String, cause: Throwable = null)
  extends Exception(message, cause) {

  def this(message: String) = this(message, null)
}

File: python/dllib/src/bigdl/dllib/utils/log4Error.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import logging

logger = logging.getLogger(__name__)


def outputUserMessage(errMsg, fixMsg=None):
    logger.error(f"\n\n****************************Usage Error************************\n" + errMsg)
    if fixMsg:
        logger.error(f"\n\n****************************How to fix*************************\n"
                     + fixMsg)
    logger.error(f"\n\n****************************Call Stack*************************")


def invalidInputError(condition, errMsg, fixMsg=None):
    if not condition:
        outputUserMessage(errMsg, fixMsg)
        raise RuntimeError(errMsg)


def invalidOperationError(condition, errMsg, fixMsg=None, cause=None):
    if not condition:
        outputUserMessage(errMsg, fixMsg)
        if cause:
            raise cause
        else:
            raise RuntimeError(errMsg)
